File: iron/bpf/src/queue_depth_dynamics.cc
// IRON: iron_headers
/*
 * Distribution A
 *
 * Approved for Public Release, Distribution Unlimited
 *
 * EdgeCT (IRON) Software Contract No.: HR0011-15-C-0097
 * DCOMP (GNAT)  Software Contract No.: HR0011-17-C-0050
 * Copyright (c) 2015-20 Raytheon BBN Technologies Corp.
 *
 * This material is based upon work supported by the Defense Advanced
 * Research Projects Agency under Contracts No. HR0011-15-C-0097 and
 * HR0011-17-C-0050. Any opinions, findings and conclusions or
 * recommendations expressed in this material are those of the author(s)
 * and do not necessarily reflect the views of the Defense Advanced
 * Research Project Agency.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/* IRON: end */

/// \file queue_depth_dynamics.cc, provides an implementation of the class for
/// tracking queue depth changes over a short time period.

#include "queue_depth_dynamics.h"

#include "iron_types.h"
#include "log.h"
#include "packet.h"
#include "unused.h"

#include <inttypes.h>

using ::iron::ConfigInfo;
using ::iron::LatencyClass;
using ::iron::Log;
using ::iron::QueueDepthDynamics;
using ::iron::Time;

/// Returns the next index in the circular buffer. Extracted to a macro to
/// avoid having to do modular arithmetic all over the place.
#define NEXT_QDD_INDEX(current_idx)             \
  ((current_idx + 1) % kNumQDDSegments)

namespace
{
  /// Class name for logging.
  const char*     UNUSED(kClassName)            = "QueueDepthDynamics";

  /// QueueDepthDynamics change rates will be the BytesPerSecond change rate
  /// over the most recent time period of approximately this amount of time.
  /// (It will include at most this amount of time, and at least this amount
  /// of time * (1 - kNumQDDSegments).)
  const Time      kChangeRateResetPeriod = Time(0.3);

  /// We will rotate the circular buffer every 1/kNumQDDSegments sections.
  /// Caching that value here saves time when we call CheckReset().
  const Time      kChangeRateRotatePeriod = kChangeRateResetPeriod.Multiply(
    (double)1/kNumQDDSegments);

  /// How long to wait after the last min_bytes_reset_period_ adjustment
  /// before we next consider increasing the value. Increasing this value will
  /// make us slower to adjust when the network patterns cause spikes and dips
  /// in queue depths. Decreasing this value could make us overshoot so that
  /// we cannot quickly react to network dynamics.
  ///
  /// If dynamic floor estimation is disabled, this value will be ignored.
  const Time      kIncrMinBytesFrequency = Time(0.05);

  /// How long to wait after the last min_bytes_reset_period_ adjustment
  /// before we next consider decreasing the value. Increasing this value will
  /// make us slower to recover after a period with a lot of spikes and dips
  /// in queue depth. Decreasing this value could make us overshoot so that
  /// we cannot properly handle spikes and dips in queue depth..
  ///
  /// If dynamic floor estimation is disabled, this value will be ignored.
  const Time      kDecrMinBytesFrequency = Time(0.3);

  /// How long to wait after the latest zombie has been added before we next
  /// consider decreasing the value. Increasing this value will make us slower
  /// to recover after a period with a lot of spikes and dips in queue
  /// depth. Decreasing this value could make us overshoot so that we cannot
  /// properly handle spikes and dips in queue depth.
  ///
  /// If dynamic floor estimation is disabled, this value will be ignored.
  const Time      kDecrMinBytesTimeSinceZombieSent = Time(2.0);

  /// How much to adjust min_bytes_reset_period_ at a time.
  ///
  /// If dynamic floor estimation is disabled, this value will be ignored.
  const Time      kMinBytesResetPeriodDelta = Time(0.2);
}

//============================================================================
QueueDepthDynamics::QueueDepthDynamics()
    : min_bytes_reset_period_(0),
      min_bytes_rotate_period_(),
      last_changed_min_bytes_period_(0),
      zombie_bytes_last_added_(0),
      dynamic_min_depths_window_(false),
      min_bytes_reset_period_lower_bound_(0),
      min_bytes_reset_period_upper_bound_(0),
      net_bytes_(),
      min_bytes_(),
      zombie_bytes_added_(),
      last_reset_net_(),
      last_reset_min_(),
      current_idx_net_(0),
      current_idx_min_(0),
      initializing_net_(true),
      net_sum_(0),
      overall_min_(std::numeric_limits<uint32_t>::max()),
      total_zombies_added_(0)
{
  for (uint8_t index = 0; index < kNumQDDSegments; index++)
  {
    net_bytes_[index]          = 0;
    zombie_bytes_added_[index] = 0;
    min_bytes_[index]          = std::numeric_limits<uint32_t>::max();
    last_reset_net_[index].GetNow();
  }
  last_reset_min_.GetNow();
  zombie_bytes_last_added_.GetNow();
  last_changed_min_bytes_period_.GetNow();
}

//============================================================================
void QueueDepthDynamics::Initialize(bool dynamic_window,
                                    double initial_window_secs,
                                    double window_lower_bound_secs,
                                    double window_upper_bound_secs)
{
  dynamic_min_depths_window_ = dynamic_window;

  min_bytes_reset_period_    = Time(initial_window_secs);
  min_bytes_rotate_period_   = min_bytes_reset_period_.Multiply(
    (double)1/kNumQDDSegments);

  min_bytes_reset_period_lower_bound_ = Time(window_lower_bound_secs);
  min_bytes_reset_period_upper_bound_ = Time(window_upper_bound_secs);
}

//============================================================================
void QueueDepthDynamics::ProcessBytesAdded(
  uint16_t bytes, LatencyClass lat, uint32_t new_depth)
{
  CheckReset();
  if (std::numeric_limits<int32_t>::max() - bytes <
      net_bytes_[current_idx_net_])
  {
    // Just leave it at the max rate to avoid overflow.
    net_bytes_[current_idx_net_] = std::numeric_limits<int32_t>::max();
  }
  else
  {
    net_bytes_[current_idx_net_] += bytes;
  }
  // This can happen if we haven't had a dequeue since the last reset, so
  // min_bytes is still at its initial maximum value.
  if (new_depth < min_bytes_[current_idx_min_])
  {
    min_bytes_[current_idx_min_] = new_depth;
  }
}

//============================================================================
void QueueDepthDynamics::ProcessZombieBytesAdded(uint16_t bytes, LatencyClass lat)
{
  if (std::numeric_limits<uint32_t>::max() - bytes <
      zombie_bytes_added_[current_idx_min_])
  {
    // Just leave it at the max bytes to avoid overflow.
    zombie_bytes_added_[current_idx_min_] =
      std::numeric_limits<uint32_t>::max();
  }
  else
  {
    zombie_bytes_added_[current_idx_min_] += bytes;
  }
  zombie_bytes_last_added_.GetNow();
}

//============================================================================
void QueueDepthDynamics::BytesRemoved(
  uint16_t bytes, LatencyClass lat, uint32_t new_depth)
{
  CheckReset();
  if (std::numeric_limits<int32_t>::min() + bytes >
      net_bytes_[current_idx_net_])
  {
    // Just leave it at the min rate to avoid overflow.
    net_bytes_[current_idx_net_] = std::numeric_limits<int32_t>::min();
  }
  else
  {
    net_bytes_[current_idx_net_] -= bytes;
  }
  if (new_depth < min_bytes_[current_idx_min_])
  {
    min_bytes_[current_idx_min_] = new_depth;
  }
}

//============================================================================
int32_t QueueDepthDynamics::GetChangeRateBytesPerSec()
{
  if (initializing_net_)
  {
    // Still starting up. Return the max allowed, since we're likely quickly
    // building up a queue during system start up.
    return std::numeric_limits<int32_t>::max();
  }
  // Compute the change rate, which is the net bytes over all segments (all
  // except current is already cached in net_sum_) divided by the time over
  // which these values are valid.
  Time time_diff = Time::Now();
  // The next index in the buffer is currently the oldest.
  Time oldest = last_reset_net_[NEXT_QDD_INDEX(current_idx_net_)];
  if (time_diff <= oldest)
  {
    // This is highly unlikely, if not impossible, since we will be in
    // "intializing_net_" period for a full kChangeRateResetPeriod. However,
    // this extra check removes any possibility of a divide by 0 error.
    return std::numeric_limits<int32_t>::max();
  }
  time_diff = time_diff.Subtract(oldest);
  double rate = static_cast<double>(net_sum_ + net_bytes_[current_idx_net_])
    / time_diff.ToDouble();
  // Truncating is fine. We're not being precise enough to care about a
  // difference of less than one byte.
  return static_cast<int32_t>(rate);
}

//============================================================================
uint32_t QueueDepthDynamics::GetMinQueueDepthBytes()
{
  // NOTE: we don't need an initialization period for min bytes the way we do
  // for change rate, because a minimum makes sense (and isn't skewed) from
  // being a minimum over a short period of time.

  // overall_min_ caches the minimum across all segments except the current
  // one.
  uint32_t min = overall_min_;
  if (min_bytes_[current_idx_min_] < overall_min_)
  {
    min = min_bytes_[current_idx_min_];
  }

  if (min < total_zombies_added_ + zombie_bytes_added_[current_idx_min_])
  {
    // This could be possible if the minimum queue depth decreased AND we
    // added zombies during this period.
    min = 0;
  }
  else
  {
    min -= total_zombies_added_;
    min -= zombie_bytes_added_[current_idx_min_];
  }
  return min;
}

//============================================================================
void QueueDepthDynamics::IncrementMinBytesResetPeriod()
{
  if (!dynamic_min_depths_window_ ||
      (Time::Now() - last_changed_min_bytes_period_ < kIncrMinBytesFrequency))
  {
    return;
  }
  if (min_bytes_reset_period_ + kMinBytesResetPeriodDelta >=
      min_bytes_reset_period_upper_bound_)
  {
    min_bytes_reset_period_ = min_bytes_reset_period_upper_bound_;
  }
  else
  {
    min_bytes_reset_period_ += kMinBytesResetPeriodDelta;
  }
  min_bytes_rotate_period_ = min_bytes_reset_period_.Multiply(
    (double)1/kNumQDDSegments);
  last_changed_min_bytes_period_.GetNow();
}

//============================================================================
void QueueDepthDynamics::DecrementMinBytesResetPeriod()
{
  if (!dynamic_min_depths_window_)
  {
    return;
  }
  Time now = Time::Now();
  if ((now - last_changed_min_bytes_period_ < kDecrMinBytesFrequency)
      || (now - zombie_bytes_last_added_ < kDecrMinBytesTimeSinceZombieSent))
  {
    return;
  }
  if (min_bytes_reset_period_ - kMinBytesResetPeriodDelta <=
      min_bytes_reset_period_lower_bound_)
  {
    min_bytes_reset_period_ = min_bytes_reset_period_lower_bound_;
  }
  else
  {
    min_bytes_reset_period_ = min_bytes_reset_period_ -
      kMinBytesResetPeriodDelta;
  }
  min_bytes_rotate_period_ = min_bytes_reset_period_.Multiply(
    (double)1/kNumQDDSegments);
  last_changed_min_bytes_period_.GetNow();
}

//============================================================================
void QueueDepthDynamics::CheckReset()
{
  Time now = Time::Now();
  // Reset and move the circular buffer along once every 1/kNumQDDSegments
  // seconds.
  if (now - last_reset_net_[current_idx_net_] > kChangeRateRotatePeriod)
  {
    if (current_idx_net_ == kNumQDDSegments - 1)
    {
      // We've now filled up an entire buffer. We have sufficient data to
      // start returning it.
      initializing_net_ = false;
    }
    uint8_t next_idx = NEXT_QDD_INDEX(current_idx_net_);
    // Update the cached sum, since we'll have a new current segment to be
    // excluded.
    net_sum_ += (net_bytes_[current_idx_net_] - net_bytes_[next_idx]);
    current_idx_net_ = next_idx;
    net_bytes_[current_idx_net_] = 0;
    last_reset_net_[current_idx_net_] = now;
  }

  if (now - last_reset_min_ > min_bytes_rotate_period_)
  {
    uint8_t next_idx = NEXT_QDD_INDEX(current_idx_min_);
    // Update the cached sum, since we'll have a new current segment to be
    // excluded.
    total_zombies_added_ +=
      (zombie_bytes_added_[current_idx_min_] - zombie_bytes_added_[next_idx]);
    current_idx_min_ = next_idx;
    min_bytes_[current_idx_min_] = std::numeric_limits<uint32_t>::max();
    zombie_bytes_added_[current_idx_min_] = 0;

    // Update the cached minimum, since we'll have a new current segment to be
    // excluded.
    overall_min_ = std::numeric_limits<uint32_t>::max();
    for (uint8_t idx = NEXT_QDD_INDEX(current_idx_min_);
         idx != current_idx_min_;
         idx = NEXT_QDD_INDEX(idx))
    {
      if (min_bytes_[idx] < overall_min_)
      {
        overall_min_ = min_bytes_[idx];
      }
    }
    last_reset_min_ = now;
  }
}

File: simulation/mobsim/QueueDepthDynamics.m
% IRON: iron_headers
%
% Distribution A
%
% Approved for Public Release, Distribution Unlimited
%
% EdgeCT (IRON) Software Contract No.: HR0011-15-C-0097
% DCOMP (GNAT)  Software Contract No.: HR0011-17-C-0050
% Copyright (c) 2015-20 Raytheon BBN Technologies Corp.
%
% This material is based upon work supported by the Defense Advanced
% Research Projects Agency under Contracts No. HR0011-15-C-0097 and
% HR0011-17-C-0050. Any opinions, findings and conclusions or
% recommendations expressed in this material are those of the author(s)
% and do not necessarily reflect the views of the Defense Advanced
% Research Project Agency.
%
% Permission is hereby granted, free of charge, to any person obtaining a copy
% of this software and associated documentation files (the "Software"), to deal
% in the Software without restriction, including without limitation the rights
% to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
% copies of the Software, and to permit persons to whom the Software is
% furnished to do so, subject to the following conditions:
%
% The above copyright notice and this permission notice shall be included in all
% copies or substantial portions of the Software.
%
% THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
% IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
% FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
% AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
% LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
% OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
% SOFTWARE.
%
% IRON: end

classdef QueueDepthDynamics < handle
    properties
        min_pkts_reset_period             %
        min_pkts_rotate_period            %
        last_changed_min_pkts_period      %
        zombie_pkts_last_added            %
        dynamic_min_depths_window         %
        min_pkts_reset_period_lower_bound %
        min_pkts_reset_period_upper_bound %
        net_pkts                          %
        min_pkts                          %
        zombie_pkts_added                 %
        last_reset_net                    %
        last_reset_min                    %
        current_idx_net                   %
        current_idx_min                   %
        initializing_net                  %
        net_sum                           %
        overall_min                       %
        total_zombies_added               %
        clock                             % pointer to the sim clock
    end
    methods
        function [ obj ] = QueueDepthDynamics ()
            
            nSegs = QDDDefs.kNumQDDSegments; %
            
            obj                                   = obj@handle();
            obj.min_pkts_reset_period             = 0;
            obj.min_pkts_rotate_period            = 0;
            obj.last_changed_min_pkts_period      = 0;
            obj.zombie_pkts_last_added            = 0;
            obj.dynamic_min_depths_window         = false;
            obj.min_pkts_reset_period_lower_bound = 0;
            obj.min_pkts_reset_period_upper_bound = 0;
            obj.net_pkts                          = zeros(nSegs,1);
            obj.min_pkts                          = double(intmax('int32')) * ones(nSegs,1);
            obj.zombie_pkts_added                 = zeros(nSegs,1);
            obj.last_reset_net                    = zeros(nSegs,1);
            obj.last_reset_min                    = 0;
            obj.current_idx_net                   = 1; % Adjusted for 1 ref
            obj.current_idx_min                   = 1; % Adjusted for 1 ref
            obj.initializing_net                  = true;
            obj.net_sum                           = 0;
            obj.overall_min                       = 0;
            obj.total_zombies_added               = 0;
        end
        
        function [] = Initialize(obj, clock, dynamic_window, ...
                initial_window_secs, window_lower_bound, ...
                window_upper_bound)
            
            obj.clock                             = clock;
            obj.dynamic_min_depths_window         = dynamic_window;
            obj.min_pkts_reset_period             = initial_window_secs;
            obj.min_pkts_reset_period_lower_bound = window_lower_bound;
            obj.min_pkts_reset_period_upper_bound = window_upper_bound;
            obj.last_changed_min_pkts_period      = obj.clock.now;
            obj.zombie_pkts_last_added            = obj.clock.now;
        end
        
        function [] = ProcessPktAdded(obj, new_depth)
            obj.CheckReset();
            if ((intmax('int32') - 1) < obj.net_pkts(obj.current_idx_net))
                % Just leave it at the max rate to avoid overflow
                obj.net_pkts(obj.current_idx_net) = intmax('int32');
            else
                obj.net_pkts(obj.current_idx_net) = ...
                    obj.net_pkts(obj.current_idx_net) + 1;
            end
            
            if (new_depth < obj.min_pkts(obj.current_idx_min))
                obj.min_pkts(obj.current_idx_min) = new_depth;
            end
        end
        
        function [] = ProcessZombiePktAdded(obj)
            if ((intmax('int32') - 1) < ...
                    obj.zombie_pkts_added_pkts(obj.curr_idx_min))
                % Just leave it at the max to avoid overflow
                obj.zombie_pkts_added(obj.curr_idx_min) = intmax('int32');
            else
                obj.zombie_pkts_added(obj.curr_idx_min) = ...
                    obj.zombie_pkts_added(obj.curr_idx_min) + 1;
            end
            obj.zombie_pkts_last_added = obj.clock.now;
        end
        
        function [] = PktRemoved(obj, new_depth)
            obj.CheckReset();
            if ((intmin('int32') + 1) > obj.net_pkts(obj.current_idx_net))
                % Just leave it at the min rate to avoid wrap around
                obj.net_pkts(obj.current_idx_net) = intmin('int32');
            else
                obj.net_pkts(obj.current_idx_net) = ...
                    obj.net_pkts(obj.current_idx_net) - 1;
            end
            
            if (new_depth < obj.min_pkts(obj.current_idx_min))
                obj.min_pkts(obj.current_idx_min) = new_depth;
            end
        end
        
        function [rate] = GetChangeRatePktsPerSec(obj)
            
            if (obj.initializing_net)
                rate = intmax('int32');
                return;
            end
            
            % Compute the change rate, which is the net pkts over all
            % segments (all except the current segment is already cached in
            % net_sum) divided by the time over whihc these values are
            % valid
            
            % The next index in the buffer is currently the oldest
            oldest = obj.last_reset_net(obj.NextQDDIndex(obj.current_idx_net));
            if (obj.clock.now <= oldest)
                % This is highly unlikely if not impossible, since we will
                % be in the "initializing_net" period for a full
                % kChangeRateResetPeriod. However, this extra check removes
                % any possibility of a divide by 0 error
                intmax('int32');
                return;
            end
            
            rate = (obj.net_sum - obj.net_pkts(obj.current_idx_net)) ...
                / (obj.clock.now - oldest);
        end
        
        function [min] = GetMinQueueDepthPkts(obj)
            min = obj.overall_min;
            if (obj.min_pkts(obj.current_idx_min) < obj.overall_min)
                min = obj.min_pkts(obj.current_idx_min);
            end
            if (min < (obj.total_zombies_added + ...
                    obj.zombie_pkts_added(obj.current_idx_min)))
                min = 0;
            else
                min = min - obj.total_zombies_added;
                min = min - obj.zombie_pkts_added(obj.current_idx_min);
            end
        end
        
        function [] = IncrementMinPktsResetPeriod(obj)
            if ((~obj.dynamic_min_depths_window) || ...
                    ((obj.clock.now - obj.last_changed_min_pkts_period) < ...
                    QDDDefs.kIncrMinPktsPeriod))
                return;
            end
            if ((obj.min_pkts_reset_period + ...
                    QDDDefs.kIncrMinPktsPeriodDelta) >= ...
                    obj.min_pkts_reset_period_upper_bound)
                obj.min_pkts_reset_period = ...
                    obj.min_pkts_reset_period_upper_bound;
            else
                obj.min_pkts_reset_period = obj.min_pkts_reset_period + ...
                    QDDDefs.kMinPktsResetPeriodDelta;
            end
            obj.min_pkts_rotate_period = min_bytes_reset_period ...
                / QDDDefs.kNumQDDSegments;
            obj.last_changed_min_pkts_period = obj.clock.now;
        end
        
        function [] = DecrementMinPktsResetPeriod(obj)
            if (~obj.dynamic_min_depths_window)
                return;
            end
            if (((obj.clock.now - obj.last_changed_min_pkts_period) < ...
                    QDDDefs.kDecrMinPktsPeriod) || ...
                    ((obj.clock.now - obj.zombie_pkts_last_added) < ...
                    QDDDefs.kDecrMinPktsTimeSinceZombieSent))
                return;
            end
            
            if ((obj.min_pkts_reset_period - ...
                    QDDDefs.kMinPktsResetPeriodDelta) <= ...
                    obj.min_pkts_reset_period_lower_bound)
                obj.min_pkts_reset_period = obj.min_pkts_reset_period_lower_bound;
            else
                obj.min_pkts_reset_period = obj.min_pkts_reset_period - ...
                    QDDDefs.kMinPktsResetPeriodDelta;
            end
            
            obj.min_pkts_rotate_period = min_bytes_reset_period ...
                / QDDDefs.kNumQDDSegments;
            obj.last_changed_min_pkts_period = obj.clock.now;
        end
        
        function [] = CheckReset(obj)
            % Reset and move the circular buffer along once every
            % 1/kNumQDDSegments seconds
            
            if ((obj.clock.now - obj.last_reset_net(obj.current_idx_net)) > ...
                    QDDDefs.kChangeRateRotatePeriod)
                if (obj.current_idx_net == QDDDefs.kNumQDDSegments)
                    % We've now filled an entire buffer. We have sufficient
                    % data to start returning it
                    obj.initializing_net = false;
                end
                
                next_idx = obj.NextQDDIndex(obj.current_idx_net);
                
                obj.net_sum = obj.net_sum + obj.net_pkts(obj.current_idx_net) - ...
                    obj.net_pkts(next_idx);
                obj.current_idx_net = next_idx;
                
                obj.net_pkts(obj.current_idx_net) = 0;
                obj.last_reset_net(obj.current_idx_net) = obj.clock.now;
            end
            
            if ((obj.clock.now - obj.last_reset_min) > ...
                    obj.min_pkts_rotate_period)
                next_idx = obj.NextQDDIndex(obj.current_idx_min);
                % Update the cached sum, since we'll have a new current
                % segment to be excluded
                obj.total_zombies_added = ...
                    obj.total_zombies_added + ...
                    obj.zombie_pkts_added(obj.current_idx_min) - ...
                    obj.zombie_pkts_added(next_idx);
                
                obj.current_idx_min = next_idx;
                obj.min_pkts(obj.current_idx_min) = intmax('int32');
                obj.zombie_pkts_added(obj.current_idx_min) = 0;
                
                % Update the cached minimum, since we'll have a new current
                % segment to be excluded
                
                obj.overall_min = intmax('int32');
                % Loop goes here
                idx = obj.NextQDDIndex(obj.current_idx_min);
                while (idx ~= obj.current_idx_min)
                    if (obj.min_pkts(idx) < obj.overall_min)
                        obj.overall_min = obj.min_pkts(idx);
                    end
                    idx = obj.NextQDDIndex(idx);
                end
                
                obj.last_reset_min = obj.clock.now;
            end
        end
        
        function [next] = NextQDDIndex(~,arg)
            next = mod(arg,QDDDefs.kNumQDDSegments) + 1;
        end
        
    end
end
